lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -384,6 +384,259 @@ class DecomposeAtenReshapeOp : public OpRewritePattern<AtenReshapeOp> {
 };
 } // namespace
 
+namespace {
+// support static shape only
+class DecomposeAtenEinsumOp : public OpRewritePattern<AtenEinsumOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenEinsumOp op,
+                                PatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+    std::string equation;
+    if (!matchPattern(op.getEquation(), m_TorchConstantStr(equation))) {
+      return rewriter.notifyMatchFailure(op, "Unsupported value of equation");
+    }
+
+    SmallVector<SmallVector<char>> inputTokens;
+    SmallVector<char> resultTokens;
+    SmallVector<char> inputToken;
+    SmallVector<SmallVector<Value>> inputShapes;
+
+    size_t index = 0;
+    enum EquationVariable { kIsInput, kIsResult };
+    EquationVariable currentVariable = kIsInput;
+    while (index < equation.size()) {
```

Comment:
I would move `inputToken` inside the `while` loop since it is only used in here. I don't know what the final state of `inputToken` is after calling `std::move` on it, so having it initialized with every `while` iteration seems safer to me.

Reply:
> I would move `inputToken` inside the `while` loop since it is only used in here. I don't know what the final state of `inputToken` is after calling `std::move` on it, so having it initialized with every `while` iteration seems safer to me.

Here we are trying to split input token and result token, the state will not gonna change once it become kIsResult, so it couldn't be put in the while loop, otherwise the state will be initialize every time